### DISCUSSION TITLE
Bridge functionality using Moquette Broker?

### DISCUSSION THREAD

**Discussion by geoxanadu** [FraunhoferIOSB/FROST-Server#539] [Q&A]
I was wondering if there is a bridge funcionality in the used moquette broker (http://owntracks.org/booklet/guide/bridge/)? This might come in handy when connecting the SensorThingsServer to existing IoT platforms. If not, have you thought to substitute the moquette with i.e. mosquitto or other brokers providing this capability?

**Comment by hylkevds** [comment]
The MQTT functionality is largely separated with a clear interface, so you can implement support for a different broker. The hard part of MQTT in SensorThings API is dealing with large number of potential topics that all need to receive the same updates.

Take for instance an update to Datastream(1). Which topics need to be notified?
v1.0/Datastreams
v1.0/Datastreams(1)
v1.0/Things(1)/Datastreams
v1.0/Sensors(1)/Datastreams
v1.0/ObservedProperties(1)/Datastreams
v1.0/Observations(xxx)/Datastream (possibly millions of topics)

And each of those can have about 3 million different combinations of $select with the 10 properties of Datastream. Since $select is not part of MQTT, the MQTT broker can not do this processing for you.

That's why the broker is part of the server, so that the server can keep track of which topics actually do have a subscription on them, so it only needs to send notifications to those. Whichever bridging solution you want to use will also need to forward the active subscriptions to all servers.

We chose moquette because it's written in Java, and it is possible to a listener that receives all subscribe and unsubscribe events. I don't know if this can be done in Mosquitto.

This also explains why the # wildcard is a problem in SensorThings API. You'd end up with millions of duplicate messages due to the $select convention.

**Comment by hylkevds** [comment]
I just had another look at the bridging guide you linked, and it seems to me that this should work find already for the collection topics. If you set up a Mosquitto MQTT server, and tell it to bridge (in) the primary collections (v1.0/Datastreams, v1.0/Observations, etc) of the SensorThingsService, the Mosquitto server should subscribe itself to those topics and thus receive all updates for those topics.

Can you give that a try and see if it works?

The biggest thing missing would be the ability to subscribe to v1.0/Datastreams(*)/Observations and v1.0/FeaturesOfInterest(*)/Observations. Any other wildcard subscription would be impossible, since all Entities except for Observations have collections.
